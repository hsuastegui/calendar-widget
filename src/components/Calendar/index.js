import React, { useState, useEffect } from "react";
import PropTypes from "prop-types";
import next from "../../assets/next.svg";
import previous from "../../assets/previous.svg";
import {
  getRows,
  getNextMonth,
  getPrevMonth,
  getCurrentMonth,
  getFormattedDate,
  getToday,
  getUpdatedDate,
  format,
} from "../../utils";
import {
  Container,
  Day,
  Bubble,
  Row,
  WeekDays,
  WeekDay,
  Header,
  Button,
} from "./style";

const week = ["Mo", "Tu", "We", "Th", "Fr", "Sa", "Su"];

function Calendar({ open, onSelect, selected }) {
  const [currentMonth, setCurrentMonth] = useState(getCurrentMonth());
  const [selectedDate, setSelectedDate] = useState("");
  const today = getToday();
  const currentDate = getFormattedDate(currentMonth);
  const year = currentDate.year();
  const month = currentDate.format("MMMM");

  useEffect(() => {
    if (selected === "") {
      setSelectedDate("");
      setCurrentMonth(getCurrentMonth());
    }
  }, [selected]);

  const handlePrev = () => {
    setCurrentMonth(getPrevMonth(currentDate));
  };

  const handleNext = () => {
    setCurrentMonth(getNextMonth(currentDate));
  };

  const handleClick = (d) => {
    const date = getUpdatedDate(currentDate, d);
    setSelectedDate(date.format(format));
    onSelect(date.format("DD/MM/YYYY"));
  };

  return (
    <Container open={open}>
      <Header>
        <Button prev onClick={handlePrev}>
          <img src={previous} alt="previous" />
        </Button>
        <span>
          {month} {year}
        </span>
        <Button next onClick={handleNext}>
          <img src={next} alt="next" />
        </Button>
      </Header>
      <WeekDays>
        {week.map((w) => (
          <WeekDay key={w}>{w}</WeekDay>
        ))}
      </WeekDays>
      {getRows(currentDate).map((r) => (
        <Row key={r.join("")}>
          {r.map((d, i) => (
            <Day key={i}>
              {d !== "" && (
                <Bubble
                  onClick={() => {
                    handleClick(d);
                  }}
                  current={today === `${year}-${month}-${d}`}
                  selected={selectedDate === `${year}-${month}-${d}`}
                >
                  {d}
                </Bubble>
              )}
            </Day>
          ))}
        </Row>
      ))}
    </Container>
  );
}

Calendar.defaultProps = {
  open: false,
  onSelect: () => {},
  selected: "",
};

Calendar.propTypes = {
  open: PropTypes.bool,
  onSelect: PropTypes.func,
  selected: PropTypes.string,
};
export default Calendar;
