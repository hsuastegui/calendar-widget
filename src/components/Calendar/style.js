import styled from "styled-components";
import {
  veryLightBlue,
  brightBlue,
  darkishBlue,
  white,
} from "../../styles/colors";
import { Roboto } from "../../styles/fonts";

export const Container = styled.div`
  width: 392px;
  border-radius: 4px;
  box-shadow: 0 0 32px 0 rgba(0, 53, 124, 0.25);
  border: solid 2px ${darkishBlue};
  background-color: ${white};
  position: absolute;
  top: 46px;
  left: -2px;
  box-sizing: border-box;
  transition: all ease 0.3s;
  overflow: hidden;
  max-height: ${({ open }) => (open ? "500px" : 0)};
  opacity: ${({ open }) => (open ? 1 : 0)};
`;

export const Header = styled.div`
  font-family: ${Roboto};
  font-size: 16px;
  font-weight: 700;
  line-height: 1.5;
  color: ${darkishBlue};
  display: flex;
  justify-content: space-between;
  align-items: center;
  border-bottom: 1px solid rgba(0, 53, 124, 0.25);
`;

export const Button = styled.button`
  background: none;
  border: none;
  cursor: pointer;
  height: 58px;
  width: 42px;
  border-right: ${({ prev }) =>
    prev ? " solid 1px rgba(0, 53, 124, 0.25)" : "none"};
  border-left: ${({ next }) =>
    next ? " solid 1px rgba(0, 53, 124, 0.25)" : "none"};
`;

export const WeekDays = styled.div`
  display: flex;
  border-bottom: solid 1px rgba(0, 53, 124, 0.25);
  padding: 16px 0;
`;

export const WeekDay = styled.span`
  font-family: ${Roboto};
  font-size: 12px;
  font-weight: 700;
  line-height: 1.33;
  color: ${darkishBlue};
  flex: 1 1 14.28%;
  text-align: center;
`;

export const Row = styled.div`
  display: flex;
  border-bottom: solid 1px rgba(0, 53, 124, 0.25);
  padding: 8px 0;
`;

export const Day = styled.div`
  flex: 0 1 14.28%;
`;

export const Bubble = styled.span`
  display: block;
  margin: 0 auto;
  border-radius: 50%;
  cursor: pointer;
  font-family: ${Roboto};
  font-size: 12px;
  line-height: 32px;
  width: 32px;
  text-align: center;
  font-weight: ${({ current, selected }) => (current || selected ? 700 : 400)};
  color: ${({ selected }) => (selected ? white : darkishBlue)};
  background: ${({ selected }) => (selected ? brightBlue : white)};

  &:hover {
    background: ${({ selected }) => (selected ? brightBlue : veryLightBlue)};
    font-weight: 700;
  }
`;
