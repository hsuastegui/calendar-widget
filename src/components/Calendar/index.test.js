import React from "react";
import { shallow } from "enzyme";
import Calendar from "./index";
import { Header, Button, Bubble } from "./style";
jest.mock("../../utils");

describe("Calendar", () => {
  const onSelect = jest.fn();
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<Calendar open onSelect={onSelect} />);
  });
  it("renders", () => {
    expect(wrapper).toMatchSnapshot();
  });
  it("displays previous month", () => {
    wrapper.find(Button).first().simulate("click");
    expect(wrapper.find(Header).text()).toEqual("May 2020");
  });
  it("displays next month", () => {
    wrapper.find(Button).last().simulate("click");
    expect(wrapper.find(Header).text()).toEqual("July 2020");
  });
  it("selects date", () => {
    wrapper.find(Bubble).at(12).simulate("click");
    expect(onSelect).toHaveBeenCalledWith("13/06/2020");
  });
});
