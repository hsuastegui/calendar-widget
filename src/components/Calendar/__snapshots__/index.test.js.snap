// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calendar renders 1`] = `
<styled.div
  open={true}
>
  <styled.div>
    <styled.button
      onClick={[Function]}
      prev={true}
    >
      <img
        alt="previous"
        src="previous.svg"
      />
    </styled.button>
    <span>
      June
       
      2020
    </span>
    <styled.button
      next={true}
      onClick={[Function]}
    >
      <img
        alt="next"
        src="next.svg"
      />
    </styled.button>
  </styled.div>
  <styled.div>
    <styled.span
      key="Mo"
    >
      Mo
    </styled.span>
    <styled.span
      key="Tu"
    >
      Tu
    </styled.span>
    <styled.span
      key="We"
    >
      We
    </styled.span>
    <styled.span
      key="Th"
    >
      Th
    </styled.span>
    <styled.span
      key="Fr"
    >
      Fr
    </styled.span>
    <styled.span
      key="Sa"
    >
      Sa
    </styled.span>
    <styled.span
      key="Su"
    >
      Su
    </styled.span>
  </styled.div>
  <styled.div
    key="1234567"
  >
    <styled.div
      key="0"
    >
      <styled.span
        current={false}
        onClick={[Function]}
        selected={false}
      >
        1
      </styled.span>
    </styled.div>
    <styled.div
      key="1"
    >
      <styled.span
        current={false}
        onClick={[Function]}
        selected={false}
      >
        2
      </styled.span>
    </styled.div>
    <styled.div
      key="2"
    >
      <styled.span
        current={false}
        onClick={[Function]}
        selected={false}
      >
        3
      </styled.span>
    </styled.div>
    <styled.div
      key="3"
    >
      <styled.span
        current={false}
        onClick={[Function]}
        selected={false}
      >
        4
      </styled.span>
    </styled.div>
    <styled.div
      key="4"
    >
      <styled.span
        current={false}
        onClick={[Function]}
        selected={false}
      >
        5
      </styled.span>
    </styled.div>
    <styled.div
      key="5"
    >
      <styled.span
        current={false}
        onClick={[Function]}
        selected={false}
      >
        6
      </styled.span>
    </styled.div>
    <styled.div
      key="6"
    >
      <styled.span
        current={false}
        onClick={[Function]}
        selected={false}
      >
        7
      </styled.span>
    </styled.div>
  </styled.div>
  <styled.div
    key="891011121314"
  >
    <styled.div
      key="0"
    >
      <styled.span
        current={false}
        onClick={[Function]}
        selected={false}
      >
        8
      </styled.span>
    </styled.div>
    <styled.div
      key="1"
    >
      <styled.span
        current={false}
        onClick={[Function]}
        selected={false}
      >
        9
      </styled.span>
    </styled.div>
    <styled.div
      key="2"
    >
      <styled.span
        current={false}
        onClick={[Function]}
        selected={false}
      >
        10
      </styled.span>
    </styled.div>
    <styled.div
      key="3"
    >
      <styled.span
        current={false}
        onClick={[Function]}
        selected={false}
      >
        11
      </styled.span>
    </styled.div>
    <styled.div
      key="4"
    >
      <styled.span
        current={false}
        onClick={[Function]}
        selected={false}
      >
        12
      </styled.span>
    </styled.div>
    <styled.div
      key="5"
    >
      <styled.span
        current={false}
        onClick={[Function]}
        selected={false}
      >
        13
      </styled.span>
    </styled.div>
    <styled.div
      key="6"
    >
      <styled.span
        current={false}
        onClick={[Function]}
        selected={false}
      >
        14
      </styled.span>
    </styled.div>
  </styled.div>
  <styled.div
    key="15161718192021"
  >
    <styled.div
      key="0"
    >
      <styled.span
        current={false}
        onClick={[Function]}
        selected={false}
      >
        15
      </styled.span>
    </styled.div>
    <styled.div
      key="1"
    >
      <styled.span
        current={false}
        onClick={[Function]}
        selected={false}
      >
        16
      </styled.span>
    </styled.div>
    <styled.div
      key="2"
    >
      <styled.span
        current={false}
        onClick={[Function]}
        selected={false}
      >
        17
      </styled.span>
    </styled.div>
    <styled.div
      key="3"
    >
      <styled.span
        current={false}
        onClick={[Function]}
        selected={false}
      >
        18
      </styled.span>
    </styled.div>
    <styled.div
      key="4"
    >
      <styled.span
        current={false}
        onClick={[Function]}
        selected={false}
      >
        19
      </styled.span>
    </styled.div>
    <styled.div
      key="5"
    >
      <styled.span
        current={false}
        onClick={[Function]}
        selected={false}
      >
        20
      </styled.span>
    </styled.div>
    <styled.div
      key="6"
    >
      <styled.span
        current={false}
        onClick={[Function]}
        selected={false}
      >
        21
      </styled.span>
    </styled.div>
  </styled.div>
  <styled.div
    key="22232425262728"
  >
    <styled.div
      key="0"
    >
      <styled.span
        current={false}
        onClick={[Function]}
        selected={false}
      >
        22
      </styled.span>
    </styled.div>
    <styled.div
      key="1"
    >
      <styled.span
        current={false}
        onClick={[Function]}
        selected={false}
      >
        23
      </styled.span>
    </styled.div>
    <styled.div
      key="2"
    >
      <styled.span
        current={false}
        onClick={[Function]}
        selected={false}
      >
        24
      </styled.span>
    </styled.div>
    <styled.div
      key="3"
    >
      <styled.span
        current={false}
        onClick={[Function]}
        selected={false}
      >
        25
      </styled.span>
    </styled.div>
    <styled.div
      key="4"
    >
      <styled.span
        current={false}
        onClick={[Function]}
        selected={false}
      >
        26
      </styled.span>
    </styled.div>
    <styled.div
      key="5"
    >
      <styled.span
        current={false}
        onClick={[Function]}
        selected={false}
      >
        27
      </styled.span>
    </styled.div>
    <styled.div
      key="6"
    >
      <styled.span
        current={true}
        onClick={[Function]}
        selected={false}
      >
        28
      </styled.span>
    </styled.div>
  </styled.div>
  <styled.div
    key="2930"
  >
    <styled.div
      key="0"
    >
      <styled.span
        current={false}
        onClick={[Function]}
        selected={false}
      >
        29
      </styled.span>
    </styled.div>
    <styled.div
      key="1"
    >
      <styled.span
        current={false}
        onClick={[Function]}
        selected={false}
      >
        30
      </styled.span>
    </styled.div>
  </styled.div>
</styled.div>
`;
